object-store/src/main/java/org/hypertrace/config/objectstore/IdentifiedObjectStoreWithFilter.java

```diff
@@ -0,0 +1,65 @@
+package org.hypertrace.config.objectstore;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.hypertrace.config.service.change.event.api.ConfigChangeEventGenerator;
+import org.hypertrace.config.service.v1.ConfigServiceGrpc;
+import org.hypertrace.core.grpcutils.context.RequestContext;
+
+/**
+ * IdentifiedObjectStoreWithFilter is an abstraction over {@link IdentifiedObjectStore} to allow
+ * filtering on fetched configs
+ *
+ * @param <T> Config Object
+ * @param <F> Filter
+ */
+public abstract class IdentifiedObjectStoreWithFilter<T, F> extends IdentifiedObjectStore<T> {
+  protected IdentifiedObjectStoreWithFilter(
+      ConfigServiceGrpc.ConfigServiceBlockingStub configServiceBlockingStub,
+      String resourceNamespace,
+      String resourceName,
+      ConfigChangeEventGenerator configChangeEventGenerator) {
+    super(configServiceBlockingStub, resourceNamespace, resourceName, configChangeEventGenerator);
+  }
+
+  public List<T> getAllFilteredConfigs(RequestContext context, F filter) {
+    Stream<T> configObjectsStream = getAllObjects(context).stream().map(ConfigObject::getData);
+    if (!isFilterEmpty(filter)) {
+      configObjectsStream =
+          configObjectsStream
+              .map(configObject -> filterConfig(configObject, filter))
+              .filter(Optional::isPresent)
+              .map(Optional::get);
+    }
+    return configObjectsStream.collect(Collectors.toUnmodifiableList());
+  }
+
+  public Optional<T> getFilteredConfig(RequestContext context, String contextId, F filter) {
+    Optional<T> fetchedConfig = getData(context, contextId);
+    if (!isFilterEmpty(filter)) {
+      return fetchedConfig.flatMap(config -> filterConfig(config, filter));
+    }
+    return fetchedConfig;
+  }
+
+  /**
+   * Method to obtain filtered config A config object can consist of multiple configs, e.g. array,
+   * which can be partially filtered and returned. In cases where filter applies to entire config
+   * object, Optional.empty() is returned
+   *
+   * @param config
+   * @param filter
+   * @return
+   */
+  protected abstract Optional<T> filterConfig(T config, F filter);
+
+  /**
+   * Method to check if any filter confitions exist at all
+   *
+   * @param filter
+   * @return
+   */
+  protected abstract boolean isFilterEmpty(F filter);
```

Comment:
Curious if we really need/want this. Usually, a caller should be using `hasFilter` style checks for presence detection, rather than handing off empty filters.

Reply:
this can be a simple `hasCheck` too.. I didn't want the caller to decide between calling the filter api or regular one.. also, if filter isn't present, no point iterating thru the stream..

Reply:
the hasCheck won't be possible since we've already had to call getFilter to get here and pass in the arg. If we don't want the caller to decide, we have a few options, none great:
1. remove any distinction on presence. In other words, an empty filter must be equivalent to an unset filter. This is what is being done here I believe.
2. pass in more context to this argument (e.g. passing in the request itself rather than - or instead of - just the filter allows the hasCheck). Downside here is that we typically have one store per resource but multiple methods (and thus multiple request shapes where that `hasCheck` might live.
3. skip the optimization (don't see a real reason to do this, at worst we could default it to false and allow opting in with an implementation if desired)

I guess after listing them out, callers that do have a distinction on presence can always opt out by putting their own check in front for example, so no real change requested here.

Reply:
hmm.. removing